libminifi/src/utils/Id.cpp

```diff
@@ -193,8 +190,8 @@ IdGenerator::~IdGenerator() = default;
 
 uint64_t IdGenerator::getDeviceSegmentFromString(const std::string& str, int numBits) const {
   uint64_t deviceSegment = 0;
-  for (size_t i = 0; i < str.length(); i++) {
-    unsigned char c = toupper(str[i]);
+  for (auto ch : str) {
+    auto c = gsl::narrow<unsigned char>(std::toupper(gsl::narrow<unsigned char>(ch)));
```

Comment:
is `gsl::narrow` the right choice here? `StringUtils` uses implicit conversions ([cppreference](https://en.cppreference.com/w/cpp/string/byte/toupper))

Reply:
the discussion mentions either `static_cast` or `gsl::narrow_cast`, `gsl::narrow` is different as it can throw, is that intentional?

Reply:
@fgerlits suggested `gsl::narrow` originally. It shouldn't matter, since this function is working with a restricted set of characters. I suggested `gsl::narrow_cast` or `static_cast`, but since it hex digits only, @fgerlits commented that terminating is OK in that case. I don't think we should enforce that here, but I don't care that much.